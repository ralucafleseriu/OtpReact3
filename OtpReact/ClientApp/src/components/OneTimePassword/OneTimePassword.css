.otp {
  padding: 1rem;
}
input {
  margin-right: 1rem;
}
section {
  height: 4rem;
}