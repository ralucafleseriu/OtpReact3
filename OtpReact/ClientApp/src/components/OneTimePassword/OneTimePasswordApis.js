export async function validatePassword(userId, password) {
  const response = await fetch('onetimepassword/validate/' + password, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(userId)
  });
  return await response.json();
}

export async function generatePassword() {
  const response = await fetch('onetimepassword');
  return await response.json();
}