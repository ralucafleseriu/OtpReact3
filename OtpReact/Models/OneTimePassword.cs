using System;

namespace OtpReact.Models
{
    public class OneTimePassword
    {
        public DateTime ExpirationDateTime { get; set; }

        public Guid UserId { get; set; }

        public string Password { get; set; }
        public OneTimePassword(Guid userId, string password, DateTime expirationDateTime)
        {
            UserId = userId;
            Password = password;
            ExpirationDateTime = expirationDateTime;
        }
    }
}
