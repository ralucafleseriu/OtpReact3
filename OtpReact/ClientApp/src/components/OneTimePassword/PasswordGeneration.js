import React, { useState, useEffect } from 'react';
import './OneTimePassword.css';
import PropTypes from 'prop-types';

const PasswordGeneration = ({ password, expirationDateTime, onGeneratePassword }) => {

  const [counter, setCounter] = useState();
  const [myInterval, setMyInterval] = useState();

  useEffect(() => {
    clearInterval(myInterval);
    setMyInterval()

    if (!expirationDateTime) return;

    setCounter(30);
    const id = setInterval(() => {
      if (expirationDateTime) {
        var c = Math.floor((expirationDateTime - new Date()) / 1000);
        setCounter(c)
      }
    }, 1000);
    setMyInterval(id);

    return () => {
      clearInterval(myInterval);
      setMyInterval()
    }
    // eslint-disable-next-line 
  }, [expirationDateTime]);

  function getPasswordStatus() {
    if (!counter) return null;
    if (counter <= 0) {
      return "Password expired.";
    }
    return `Valid for ${counter} seconds.`;
  }

  return (
    <div>
      <input value={password} readOnly />
      <button
        disabled={!!counter && counter > 0}
        onClick={() => onGeneratePassword()}
      >Generate one time password</button>
      <p>{getPasswordStatus()}</p>

    </div>
  );
}
PasswordGeneration.propTypes = {
  password: PropTypes.string,
  expirationDateTime: PropTypes.instanceOf(Date),
  onGeneratePassword: PropTypes.func
}
export default PasswordGeneration;
