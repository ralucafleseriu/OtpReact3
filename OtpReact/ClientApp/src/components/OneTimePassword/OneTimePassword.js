import React, { useState, useEffect } from 'react';
import './OneTimePassword.css';
import { validatePassword, generatePassword } from './OneTimePasswordApis';

const OneTimePassword = () => {

  const [password, setPassword] = useState("");
  const [userId, setUserId] = useState("");
  const [expirationDateTime, setExpirationDateTime] = useState();
  const [counter, setCounter] = useState();
  const [myInterval, setMyInterval] = useState();

  async function handleGeneratePassword() {
    const data = await generatePassword();
    setPassword(data.password);
    setUserId(data.userId);
    setExpirationDateTime(new Date(data.expirationDateTime));

  }

  async function handleValidatePassword(password) {
    const data = await validatePassword(userId, password);
    alert(data ? "Password valid!" : "Password expired... :( ");
    if (!data) {
      setPassword();
      setCounter()
      setExpirationDateTime();
    }
  }

  useEffect(() => {
    clearInterval(myInterval);
    setMyInterval()

    if (!expirationDateTime) return;

    setCounter(30);
    const id = setInterval(() => {
      if (expirationDateTime) {
        var c = Math.floor((expirationDateTime - new Date()) / 1000);
        setCounter(c)
      }
    }, 1000);
    setMyInterval(id);

    return () => {
      clearInterval(myInterval);
      setMyInterval()
    }
    // eslint-disable-next-line 
  }, [expirationDateTime]);

  function getPasswordStatus() {
    if (!counter) return null;
    if (counter <= 0) {
      return "Password expired.";
    }
    return `Valid for ${counter} seconds.`;
  }

  return (
    <div className="otp">
      <h1>One Time Password Generator</h1>
      <section>
        <input value={password} readOnly />
        <button
          disabled={!!counter && counter > 0}
          onClick={() => handleGeneratePassword()}
        >Generate one time password </button>
        <p>{getPasswordStatus()}</p>

      </section>
      <button
        onClick={() => handleValidatePassword(password)}
      >Validate password</button>
    </div >
  );
}

export default OneTimePassword;
