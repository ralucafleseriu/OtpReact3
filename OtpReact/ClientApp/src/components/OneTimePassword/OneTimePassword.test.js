import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import OneTimePassword from './OneTimePassword';

function renderOTP(args) {
  const defaultProps = {}
  const props = { ...defaultProps, ...args }
  return shallow(<OneTimePassword {...props} />);
}

it('renders without crashing', () => {
  shallow(<OneTimePassword />);
});
it('OTP renders correctly', () => {
  const tree = renderer
    .create(<OneTimePassword />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders fields correctly', () => {
  const wrapper = renderOTP();
  expect(wrapper.find("input").length).toBe(1);
  expect(wrapper.find("button").length).toBe(2);
});