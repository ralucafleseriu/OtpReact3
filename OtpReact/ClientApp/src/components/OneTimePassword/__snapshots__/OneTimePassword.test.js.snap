// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OTP renders correctly 1`] = `
<div
  className="otp"
>
  <h1>
    One Time Password Generator
  </h1>
  <section>
    <input
      readOnly={true}
      value=""
    />
    <button
      disabled={false}
      onClick={[Function]}
    >
      Generate one time password 
    </button>
    <p />
  </section>
  <button
    onClick={[Function]}
  >
    Validate password
  </button>
</div>
`;
